node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -0,0 +1,629 @@
+	Backed,
+}
+
+struct CandidateEntry {
+	candidate_hash: CandidateHash,
+	relay_parent: Hash,
+	candidate: ProspectiveCandidate,
+	state: CandidateState,
+	erasure_root: Hash,
+}
+
+/// The scope of a [`FragmentTree`].
+pub(crate) struct Scope {
+	para: ParaId,
+	relay_parent: RelayChainBlockInfo,
+	ancestors: BTreeMap<BlockNumber, RelayChainBlockInfo>,
+	ancestors_by_hash: HashMap<Hash, RelayChainBlockInfo>,
+	base_constraints: Constraints,
+	max_depth: usize,
+}
+
+/// An error variant indicating that ancestors provided to a scope
+/// had unexpected order.
+#[derive(Debug)]
+pub struct UnexpectedAncestor;
+
+impl Scope {
+	/// Define a new [`Scope`].
+	///
+	/// All arguments are straightforward except the ancestors.
+	///
+	/// Ancestors should be in reverse order, starting with the parent
+	/// of the `relay_parent`, and proceeding backwards in block number
+	/// increments of 1. Ancestors not following these conditions will be
+	/// rejected.
+	///
+	/// This function will only consume ancestors up to the `min_relay_parent_number` of
+	/// the `base_constraints`.
+	///
+	/// Only ancestors whose children have the same session as the relay-parent's
+	/// children should be provided.
+	///
+	/// It is allowed to provide zero ancestors.
+	pub fn with_ancestors(
+		para: ParaId,
+		relay_parent: RelayChainBlockInfo,
+		base_constraints: Constraints,
+		max_depth: usize,
+		ancestors: impl IntoIterator<Item = RelayChainBlockInfo>,
+	) -> Result<Self, UnexpectedAncestor> {
+		let mut ancestors_map = BTreeMap::new();
+		let mut ancestors_by_hash = HashMap::new();
+		{
+			let mut prev = relay_parent.number;
+			for ancestor in ancestors {
+				if prev == 0 {
+					return Err(UnexpectedAncestor)
+				} else if ancestor.number != prev - 1 {
+					return Err(UnexpectedAncestor)
+				} else if prev == base_constraints.min_relay_parent_number {
+					break
+				} else {
+					prev = ancestor.number;
+					ancestors_by_hash.insert(ancestor.hash, ancestor.clone());
+					ancestors_map.insert(ancestor.number, ancestor);
+				}
+			}
+		}
+
+		Ok(Scope {
+			para,
+			relay_parent,
+			base_constraints,
+			max_depth,
+			ancestors: ancestors_map,
+			ancestors_by_hash,
+		})
+	}
+
+	fn earliest_relay_parent(&self) -> RelayChainBlockInfo {
+		self.ancestors
+			.iter()
+			.next()
+			.map(|(_, v)| v.clone())
+			.unwrap_or_else(|| self.relay_parent.clone())
+	}
+
+	fn ancestor_by_hash(&self, hash: &Hash) -> Option<RelayChainBlockInfo> {
+		if hash == &self.relay_parent.hash {
+			return Some(self.relay_parent.clone())
+		}
+
+		self.ancestors_by_hash.get(hash).map(|info| info.clone())
+	}
+}
+
+// We use indices into a flat vector to refer to nodes in the tree.
+// Every tree also has an implicit root.
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum NodePointer {
+	Root,
+	Storage(usize),
+}
+
+/// This is a tree of candidates based on some underlying storage of candidates
+/// and a scope.
+pub(crate) struct FragmentTree {
```

Comment:
> back multiple candidates of a single parachain per relay chain block - right

We don't require the relay-parent to advance but this doesn't require #4951 .

Reply:
Anyway, it would always be a tree even if we hadn't relaxed that requirement. this just complicates things because it allows for cycles.

I think you're confusing the things _we care about_ in the tree vs. the tree itself. When building a new relay-chain block, you only really care about depth=1, the stuff right by the root. But when backing/seconding new parachain blocks, you care about the stuff towards the leaves.